README.md
npstoret2R
==========

NPSTORET to R importing


NEWS
npstoret2R 0.0.0.9015
===================

DEVELOPMENT INCREMENTAL RELEASE 15

* Modified SQLquery_NPSTORET_WQProjStnCritStds SQL

* Updated documentation


npstoret2R 0.0.0.9014
===================

DEVELOPMENT INCREMENTAL RELEASE 14

* Added trim(), trim.leading() & trim.trailing() functions

* Updated documentation


npstoret2R 0.0.0.9013
===================

DEVELOPMENT INCREMENTAL RELEASE 13

* Add addKey() & convertToDataTable() functions

* Rename R_helper_functions to R_utility_functions and update documentation

* Added data.table (>=1.9.4) and plyr (>=1.8.1) dependencies


npstoret2R 0.0.0.9012
===================

DEVELOPMENT INCREMENTAL RELEASE 12

* Update SQL queries to pull directly from NPSTORET (version 1.86) and definition tables vs linking tables in Access (c:\NPSTORET\LinkTabs\NPSTORET_defTab.mdb)

* Revise Characteristics query to remove NPSTORET fields which no longer exist in tblCharacteristics (TSRCHAR_ORG_ID, TSRBioPt_ORG_ID, TSRSubTax_ORG_ID)

* Update pkgTest() & R_helper_functions documentation


npstoret2R 0.0.0.9011
===================

DEVELOPMENT INCREMENTAL RELEASE 11

* Revert package R dependency to version 3.1.0

* Specify package dependencies for RODBC(>=1.3-10), sqldf(>=0.4-10), tools(>=) and knitr(>=1.8)


npstoret2R 0.0.0.9010
===================

DEVELOPMENT INCREMENTAL RELEASE 10

* Installation test decreasing R dependency to version 3.0.0


npstoret2R 0.0.0.9009
===================

DEVELOPMENT INCREMENTAL RELEASE 9

* Continued code cleanup

* Add park comparator for result lookups


npstoret2R 0.0.0.9008
===================

DEVELOPMENT INCREMENTAL RELEASE 8

* Continued code cleanup

* Fixed dfIndepStds 'IsDependent' not found/unused argument errors

* Fixed getDepCharResult() sqldf could not find function error


npstoret2R 0.0.0.9007
===================

DEVELOPMENT INCREMENTAL RELEASE 7

* Continued code cleanup

* Adjust db connect environment


npstoret2R 0.0.0.9006
===================

DEVELOPMENT INCREMENTAL RELEASE 6

* Continued code cleanup

* Partial fix for S4 db class connect method


npstoret2R 0.0.0.9005
===================

DEVELOPMENT INCREMENTAL RELEASE 5

* Continued code cleanup

* Fixed S4 db class method export


npstoret2R 0.0.0.9004
===================

DEVELOPMENT INCREMENTAL RELEASE 4

* Continued code cleanup

* Revised app connection class to S4 and renamed to db class


npstoret2R 0.0.0.9003
===================

DEVELOPMENT INCREMENTAL RELEASE 3

* Continued code cleanup

* Updated initializeEnvironment to drop unnecessary script calls


npstoret2R 0.0.0.9002
===================

DEVELOPMENT INCREMENTAL RELEASE 2

* Continued code cleanup & documentation fixes

* Updated SQLquery_NPSTORET for NPSTORET 1.86 removal of TSRBIO_PT table


npstoret2R 0.0.0.9001
===================

DEVELOPMENT INCREMENTAL RELEASE

* adds code vignettes for examples

* Cleans up documentation


npstoret2R 0.0.0.9000
===================

INITIAL RELEASE

* enables pulling of NPSTORET data into R for results and standards (characteristics)

* finds dependent characteristics of dependent standards and calculates standard for 
  result comparison

* compares results to standard values and assesses result compliance to standard
  (+1 = above standard, 0 = standard, -1 = below standard)

* new classes for sample, standards(independent & dependent)
  used to enable object creation and storage of data

* incorporated water quality NPSTORET import


R/R_db_class.R
############################################################
# R App Settings
############################################################
# Purpose:    Single file access to application settings class
#
# Notes:      N/A
#
# Sources:  See individual functions
#              
# Revisions:  0.1  2014-10-05  B. Campbell  initial version
#             0.2  2014-11-17  B. Campbell  updated to S4 class
#             0.3  2014-11-24  B. Campbell  fixed connect slot
#             0.4  2014-11-29  B. Campbell  updated connect documentation
# ==========================================================

# ----------------------------
#  Database S3 Class
# ----------------------------

# ----------------------------------------------------------------------
#' @title RODBC
#' @name RODBC-class
#' @details Handles ODBC database connections to R.
#'
#' @section Requirements:
#' R Libraries:
#' \itemize{
#'  \item \link[RODBC]{RODBC}
#' }
#'
#' @section References:
#'   \tabular{lll}{
#'   \tab Jan. 30, 2012 \tab Kyle Brandt \cr
#'   \tab\tab http://stackoverflow.com/questions/9067492/example-of-using-an-s3-class-in-a-s4-object \cr
#'   \tab Oct. 14, 2011 \tab Triad Sou. \cr
#'   \tab\tab http://stackoverflow.com/questions/7758748/documenting-setas-and-setoldclass-with-roxygen \cr
#'   }
#' @section Sources:
#'   \tabular{llllllll}{
#'   \tab 2014-11-24 \tab\tab B. Campbell \tab\tab 0.1 \tab\tab Initial version \cr
#'   }
#' @section Revisions:
#'   \tabular{llllllll}{
#'   \tab 0.1   \tab\tab 2014-11-24  \tab\tab BLC   \tab\tab Added to resolve S4 connect slot error \cr
#'   }
#'
#' @family RODBC
#' @exportClass RODBC
# ----------------------------------------------------------------------
setOldClass("RODBC")

# ----------------------------------------------------------------------
#' @title db Class
#' @details Creates a database object whose connection parameters are contained within the object.
#'
#' @slot dbfile         - database name & directory path
#' @slot user           - database username (for authentication)
#' @slot pwd            - database password (for authentication)
#' @slot connect        - database connection
#' 
#' @examples
#' ## Create the database object (db) and assign the NPSTORET backend filename.
#' ## Paths should be written UNIX style with / vs. \ .
#' ## So a windows file path like "C:\db\dbname.mdb" should be written "C:/db/dbname.mdb".
#' 
#' \dontrun{
#' 
#' Connect to your database via a DSN-less connection.
#' 
#'connect(mynpstoret,dbfile="C:/NPSTORET/NPSTORET_BE.MDB", user="myusername", pwd="mypwd")
#'}
#'
#' @section Requirements:
#' R Libraries:
#' \itemize{
#'  \item \link[RODBC]{RODBC}
#'  \item \link[tools]{tools}
#' }
#'
#' @section References:
#'   \tabular{lll}{
#'   \tab 2004 \tab Friedrich Leisch \cr
#'   \tab\tab http://www.r-project.org/conferences/useR-2004/Keynotes/Leisch.pdf \cr
#'   \tab May 6, 2014 \tab Rappster \cr
#'   \tab\tab http://stackoverflow.com/questions/23495627/roxygen2-s4-generic-functions-are-not-exported-unless-a-method-is-also-defined \cr
#'   }
#' @section Sources:
#'   \tabular{llllllll}{
#'   \tab 2014-10-05 \tab\tab B. Campbell \tab\tab 0.1 \tab\tab Initial version \cr
#'   }
#' @section Revisions:
#'   \tabular{llllllll}{
#'   \tab 0.1   \tab\tab 2014-10-05  \tab\tab BLC   \tab\tab Initial version \cr
#'   \tab 0.2   \tab\tab 2014-11-07  \tab\tab BLC   \tab\tab Documentation update & removed sqlQUeriesFilepath since SQL queries are integrated \cr
#'   \tab 0.3   \tab\tab 2014-11-13  \tab\tab BLC   \tab\tab Documentation update \cr
#'   \tab 0.4   \tab\tab 2014-11-17  \tab\tab BLC   \tab\tab Converted to S4 class & renamed to db vs. app \cr
#'   \tab 0.5   \tab\tab 2014-11-23  \tab\tab BLC   \tab\tab Changed to Ref class vs S4 to resolve connect slot error \cr
#'   \tab 0.6   \tab\tab 2014-12-29  \tab\tab BLC   \tab\tab Documentation update \cr
#'   }
#'    
#' @family Application settings
#' @exportClass db
# ----------------------------------------------------------------------

# ----------------------------
# NPSTORET Database Attributes
# ----------------------------
#user, pwd, dbpath
setClass(
    # set db file path
    Class="db",
        
    # define the slots
    representation=representation(
      dbfile="character",
      user = "character",
      pwd = "character",
      connect = "RODBC"
      ),
    
    # ensure connect is valid
    prototype = prototype(
      connect = structure(list(), class="RODBC")
      )
    
    # validate if data is consistent
    # not called if an initialize function is defined!
#    validity=function(object){
#      if(!(file.exists(object@dbfile))){
#        return("Database file doesn't exist.")
#        stop ("Please check your database file location. Also ensure all \\ are replaced by //
#            since R doesn't translate the \\ in the same way Windows does.")
        
#      }
#      return(TRUE)
#    }
  )

# ----------------------------
#  Database Class Methods
# ----------------------------
# ----------------------------------------------------------------------
#' @title setDbFile
#' @details Set the database file name and path.
#'
#' @param dbObject - database object
#' @param dbfile   - database name & directory path
#'
#' @examples
#' \dontrun{
#'   #set database file & path
#'   setDbFile(db,"C:/mydatabase/dbfile.accb")
#' }
#'
#' @section Requirements:
#' R Libraries:
#' \itemize{
#'  \item \link[RODBC]{RODBC}
#'  \item \link[tools]{tools}
#' }
#'
#' @section References:
#'   \tabular{lll}{
#'   \tab 2004 \tab Friedrich Leisch \cr
#'   \tab\tab http://www.r-project.org/conferences/useR-2004/Keynotes/Leisch.pdf \cr
#'   \tab May 6, 2014 \tab Rappster \cr
#'   \tab\tab http://stackoverflow.com/questions/23495627/roxygen2-s4-generic-functions-are-not-exported-unless-a-method-is-also-defined \cr
#'   \tab March 11, 2014 \tab Hadley Wickham \cr
#'   \tab\tab https://groups.google.com/forum/#!topic/rdevtools/sq3DG0oj058 \cr
#'   }
#' @section Sources:
#'   \tabular{llllllll}{
#'   \tab 2014-11-18 \tab\tab B. Campbell \tab\tab 0.1 \tab\tab Initial version \cr
#'   }
#' @section Revisions:
#'   \tabular{llllllll}{
#'   \tab 0.1   \tab\tab 2014-11-18  \tab\tab BLC   \tab\tab Initial version \cr
#'   }
#'    
#' @family Application settings
#' @exportMethod setDbFile
# ----------------------------------------------------------------------
# setDbFile generic & method
setGeneric(name="setDbFile",
           def=function(dbObject,dbfile){
             standardGeneric("setDbFile")
           }
)
setMethod(f="setDbFile",
          signature="db",
          definition = function(dbObject, dbfile){
            dbObject@dbfile <- dbfile            
            return(dbObject)
          }
)

# ----------------------------------------------------------------------
#' @title setDbUser
#' @details Sets database username
#'
#' @param dbObject - database object
#' @param user     - database username (for authentication)
#'
#' @examples
#' \dontrun{
#'  setDbUser(db,"myusername")
#'}
#'
#' @section Requirements:
#' R Libraries:
#' \itemize{
#'  \item \link[RODBC]{RODBC}
#'  \item \link[tools]{tools}
#' }
#'
#' @section References:
#'   \tabular{lll}{
#'   \tab 2004 \tab Friedrich Leisch \cr
#'   \tab\tab http://www.r-project.org/conferences/useR-2004/Keynotes/Leisch.pdf \cr
#'   \tab May 6, 2014 \tab Rappster \cr
#'   \tab\tab http://stackoverflow.com/questions/23495627/roxygen2-s4-generic-functions-are-not-exported-unless-a-method-is-also-defined \cr
#'   \tab March 11, 2014 \tab Hadley Wickham \cr
#'   \tab\tab https://groups.google.com/forum/#!topic/rdevtools/sq3DG0oj058 \cr
#'   }
#' @section Sources:
#'   \tabular{llllllll}{
#'   \tab 2014-11-18 \tab\tab B. Campbell \tab\tab 0.1 \tab\tab Initial version \cr
#'   }
#' @section Revisions:
#'   \tabular{llllllll}{
#'   \tab 0.1   \tab\tab 2014-11-18  \tab\tab BLC   \tab\tab Initial version \cr
#'   }
#'    
#' @family Application settings
#' @exportMethod setDbUser
# ----------------------------------------------------------------------
# setDbUser generic & method
setGeneric(name="setDbUser",
           def=function(dbObject,user){
             standardGeneric("setDbUser")
           }
)
setMethod(f="setDbUser",
          signature="db",
          definition = function(dbObject, user){
            dbObject@user <- user            
            return(dbObject)
          }
)

# ----------------------------------------------------------------------
#' @title setDbPwd
#' @details Sets database user password
#'
#' @param dbObject - database object
#' @param pwd      - database password (for authentication)
#'
#' @examples
#' 
#' \dontrun{
#'  setDbPwd(db,"mypassword")
#'}
#'
#' @section Requirements:
#' R Libraries:
#' \itemize{
#'  \item \link[RODBC]{RODBC}
#'  \item \link[tools]{tools}
#' }
#'
#' @section References:
#'   \tabular{lll}{
#'   \tab 2004 \tab Friedrich Leisch \cr
#'   \tab\tab http://www.r-project.org/conferences/useR-2004/Keynotes/Leisch.pdf \cr
#'   \tab May 6, 2014 \tab Rappster \cr
#'   \tab\tab http://stackoverflow.com/questions/23495627/roxygen2-s4-generic-functions-are-not-exported-unless-a-method-is-also-defined \cr
#'   \tab March 11, 2014 \tab Hadley Wickham \cr
#'   \tab\tab https://groups.google.com/forum/#!topic/rdevtools/sq3DG0oj058 \cr
#'   }
#' @section Sources:
#'   \tabular{llllllll}{
#'   \tab 2014-11-18 \tab\tab B. Campbell \tab\tab 0.1 \tab\tab Initial version \cr
#'   }
#' @section Revisions:
#'   \tabular{llllllll}{
#'   \tab 0.1   \tab\tab 2014-11-18  \tab\tab BLC   \tab\tab Initial version \cr
#'   }
#'    
#' @family Application settings
#' @exportMethod setDbPwd
# ----------------------------------------------------------------------
# setDbPwd generic & method
setGeneric(name="setDbPwd",
           def=function(dbObject,pwd){
             standardGeneric("setDbPwd")
           }
)
setMethod(f="setDbPwd",
          signature="db",
          definition = function(dbObject, pwd){
            dbObject@pwd <- pwd            
            return(dbObject)
          }
)
# ----------------------------
# NPSTORET Database Connection
# ----------------------------
# ----------------------------------------------------------------------
#' @title connect
#' @details Creates DSN-less database RODBC connection
#'
#' @param dbObject - database object
#' @param dbfile   - database name & directory path
#' @param user     - database username (for authentication)
#' @param pwd      - database password (for authentication)
#' 
#' @examples
#' 
#' \dontrun{
#'  connect(mydb, dbfilename="C:/database/npstoret.accb", user="myusername", pwd="mypassword") 
#'}
#'
#' @section Requirements:
#' R Libraries:
#' \itemize{
#'  \item \link[RODBC]{RODBC}
#'  \item \link[tools]{tools}
#' }
#'
#' @section References:
#'   \tabular{lll}{
#'   \tab 2004 \tab Friedrich Leisch \cr
#'   \tab\tab http://www.r-project.org/conferences/useR-2004/Keynotes/Leisch.pdf \cr
#'   \tab May 6, 2014 \tab Rappster \cr
#'   \tab\tab http://stackoverflow.com/questions/23495627/roxygen2-s4-generic-functions-are-not-exported-unless-a-method-is-also-defined \cr
#'   \tab March 11, 2014 \tab Hadley Wickham \cr
#'   \tab\tab https://groups.google.com/forum/#!topic/rdevtools/sq3DG0oj058 \cr
#'   \tab Dec. 12, 2012 \tab Martin Morgan \cr
#'   \tab\tab http://stackoverflow.com/questions/13841400/use-s3-virtual-class-as-slot-of-an-s4-class-got-error-got-class-s4-should-b \cr
#'   }
#' @section Sources:
#'   \tabular{llllllll}{
#'   \tab 2014-11-18 \tab\tab B. Campbell \tab\tab 0.1 \tab\tab Initial version \cr
#'   }
#' @section Revisions:
#'   \tabular{llllllll}{
#'   \tab 0.1   \tab\tab 2014-11-18  \tab\tab BLC   \tab\tab Initial version \cr
#'   \tab 0.2   \tab\tab 2014-12-01  \tab\tab BLC   \tab\tab Fixed check file bug (use == not =) \cr
#'   \tab 0.3   \tab\tab 2014-12-29  \tab\tab BLC   \tab\tab Updated documentation \cr
#'   }
#'    
#' @family Application settings
#' @exportMethod connect
# ----------------------------------------------------------------------
# connection generic & method
setGeneric(name="connect",
           def=function(dbObject,dbfile,user,pwd){
             standardGeneric("connect")
           }
)
setMethod(f="connect",
          signature="db",
          definition = function(dbObject, dbfile, user, pwd){
            #library("tools")
            
            # check file is access app (do outside of connect.app): 
            if((file_ext(dbfile) == 'mapp') | (file_ext(dbfile) == 'accb')){

              # set values
              setDbFile(dbfile)
              setUser(user)
              setPwd(pwd)
              
              # check file is readable
              if(file.access(dbfile,mode=4) == 0){  
                dbObject@connect <- odbcConnectAccess2007(dbfile)   
                return(dbObject)
              }
            }        
          }
)